futures-util/src/future/shared.rs

```diff
@@ -70,7 +70,8 @@ impl<Fut: Future> fmt::Debug for Inner<Fut> {
 
 enum FutureOrOutput<Fut: Future> {
     Future(Fut),
-    Output(Arc<Fut::Output>),
+    Output(Fut::Output),
+    Done,
 }
 
 unsafe impl<Fut> Send for Inner<Fut>
```

Comment:
@cramertj Do we actually need the `where Fut::Output: Sync`?

Reply:
It is required if the output is going to accessed from an `Arc` (since that is `Arc<T: Send + Sync>`), but if `Arc` is removed from this type, then thread safety of the `Clone` is then left up to the user.

Reply:
`Arc` only implements both `Send` and `Sync` only for `T: Send + Sync`. So, how much can we reduce these bounds? Maybe `Fut::Output: Send` for `Send`, but do we require something for `Sync`? My thinking is that we don't actually have shared references, so no restrictions for `Fut::Ouptut` are required to make it `Sync`.